### PR TITLE
TimedObjectPool

### PR DESCRIPTION
**Pull request by anwar6953**
Implement a TimedObjectPool, which is an object pool that times out (destroys and discards the objects) after a certain amount of time of non-use.

https://issues.cask.co/browse/CDAP-4067
https://issues.cask.co/browse/TEPHRA-142

Testing on cluster currently in progress.

### CODE REVIEW

File: tephra-core/src/main/java/co/cask/tephra/distributed/TimedObjectPool.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    private long getTimestamp() {
+      return timestamp;
+    }
+  }
+
+  // holds all currently available elements, with the elements at the head
+  // being the most recently accessed
+  private Deque<TimestampedEntry> elements;
```
Comment (chtyim) [inline tephra-core/src/main/java/co/cask/tephra/distributed/TimedObjectPool.java]: final?

File: tephra-core/src/test/java/co/cask/tephra/distributed/TimedObjectPoolTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public int getId() {
+      return id;
+    }
+  }
+
+  private static final class ElementWithIdPool extends TimedObjectPool<ElementWithId, Exception> {
+
+    private int counter;
```
Comment (bdmogal) [inline tephra-core/src/test/java/co/cask/tephra/distributed/TimedObjectPoolTest.java]: final

File: tephra-core/src/main/java/co/cask/tephra/TxConstants.java [hunk 3]
```diff
@@ -239,16 +243,20 @@
       = "data.tx.client.retry.backoff.limit";
 
     /** the default tx client socket timeout in milli seconds. */
-    public static final int DEFAULT_DATA_TX_CLIENT_TIMEOUT
+    public static final int DEFAULT_DATA_TX_CLIENT_TIMEOUT_MS
       = 30 * 1000;
 
-    /** default number of pooled tx clients. */
+    /** default number of tx clients for client provider "pool". */
```
Comment (bdmogal) [inline tephra-core/src/main/java/co/cask/tephra/TxConstants.java]: nit: why quotes?

File: tephra-core/src/main/java/co/cask/tephra/distributed/TimedObjectPool.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+   *
+   * @param element the element to destroy
+   */
+  protected void destroy(T element) {
+    // by default do nothing
+  }
+
+  /**
+   * A method to recycle an existing element when it is returned to the pool.
+   * This methods ensures that the element is in a fresh state before it can
```
Comment (bdmogal) [inline tephra-core/src/main/java/co/cask/tephra/distributed/TimedObjectPool.java]: nit: method

**Comment by anwar6953** [comment]
Closing this PR; replacing with a simpler implementation of ElasticPool.java:
https://github.com/caskdata/tephra/pull/95

**Pull request closed without merging**